### PR TITLE
Test/darius' screens and others

### PR DESCRIPTION
**Pull request by userdarius**
# Pull Request: Add Unit Tests for Screens and Components; Fix Imports and Mocking Issues

## Description

This pull request adds comprehensive unit tests for several screens and components in the application, improving test coverage and ensuring code reliability. It also includes modifications to resolve import issues and enhance the test environment setup.

### Changes Include:

1. **Added Unit Tests for `CodeEntryScreen`**
   - Created tests to verify rendering and functionality.
   - Ensured proper handling of user inputs and navigation upon code submission.

2. **Modified `codeApproved` Screen**
   - Updated the screen to correctly handle `email` and `userId` properties.
   - Improved data flow and testability.

3. **Added Tests for `codeApprovedScreen`**
   - Implemented unit tests to verify correct information display and user interactions.

4. **Added Tests for `CreateNewCode` Screen**
   - Wrote tests covering input handling, code generation, and sharing functionality.
   - Mocked necessary modules and functions (`Firebase` utilities, `Clipboard`, `Share`, `Alert`).
   - Modified problematic imports to resolve test environment issues.

5. **Added Tests for `HomepageScreen`**
   - Implemented tests to ensure correct rendering.
   - Verified navigation to the Settings screen when prompted.
   - Mocked navigation functions to simulate user interactions.

6. **Added Tests for Additional Screens and Components**
   - **`SettingsScreen`**: Tested rendering and sign-out functionality, mocking Firebase auth.
   - **`SubrentScreen`**: Verified the screen renders the expected content.
   - **`SharedElementsScreen`**: Ensured the screen displays shared resources information correctly.
   - **`AdaptiveButton` Component**: Tested rendering with different props, including title, onPress function, icons, and styling.

7. **Fixed Import and Mocking Issues**
   - Resolved errors related to deprecated or extracted modules from `react-native` (e.g., `Clipboard`).
   - Updated imports to use the correct packages (e.g., `@react-native-clipboard/clipboard`).
   - Mocked native modules causing `Invariant Violation` errors in the test environment.
   - Adjusted test files to avoid mocking the entire `react-native` module, preventing unintended side effects.

8. **Improved Test Environment Setup**
   - Mocked `useNavigation` hook from `@react-navigation/native` to provide a mock navigation object in tests.
   - Ensured components depending on navigation context (e.g., `Header` component) function correctly during testing.
   - Used `jest.spyOn` to mock specific methods instead of entire modules, enhancing test reliability.

## Motivation and Context

- **Increase Test Coverage**: Enhancing test coverage ensures that components function as expected and helps catch regressions early.
- **Improve Code Quality**: Modifications improve data handling, align with best practices, and make components more testable.
- **Resolve Testing Issues**: Addressing import and mocking issues allows for smoother development and testing processes.

## How Has This Been Tested?

- All new unit tests have been run and pass successfully.
- Mocked functions and modules have been verified to work without causing errors.
- Components render correctly, and user interactions produce the expected outcomes in tests.

**Assistance Acknowledgment**: Portions of the code and certain testing strategies were developed with the help of AI assistance.

### CODE REVIEW

File: app/__tests__/CodeApprovedScreen.test.tsx [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+jest.mock("@react-navigation/native", () => {
+  const actualNav = jest.requireActual("@react-navigation/native");
+  return {
+    ...actualNav,
+    useNavigation: () => ({
+      navigate: mockNavigate,
+    }),
+    useRoute: () => ({
+      params: {
+        userId: "test-user-id",
```
Comment (mdesautel) [inline app/__tests__/CodeApprovedScreen.test.tsx:19]: Maybe you can test the edge case when no params, or "undefined" ones, are passed to the screen, to see if the app handles it gracefully and doesn't crash, etc.

File: app/__tests__/CodeApprovedScreen.test.tsx [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+);
+
+describe("CodeApprovedScreen", () => {
+  beforeEach(() => {
+    // Clear all mocks before each test
+    jest.clearAllMocks();
+  });
+
+  it("renders correctly", () => {
+    const { getByText, getByTestId } = render(
```
Comment (mdesautel) [inline app/__tests__/CodeApprovedScreen.test.tsx:43]: I think you should use the screen.getByText, screen.getByTestId methods from the screen API. Right now you render the screen and destructure its return (you assign to consts some parts of what the method returns), so it is not best practice and less readable.

File: app/__tests__/CreateNewCode.test.tsx [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+    // Press create code without entering any inputs
+    fireEvent.press(createCodeButton);
+
+    expect(alertMock).toHaveBeenCalledWith(
+      "Please enter both Residence ID and Apartment ID."
+    );
+
+    alertMock.mockRestore();
+  });
+
```
Comment (mdesautel) [inline app/__tests__/CreateNewCode.test.tsx:87]: You can also test the case where you enter only one of the two inputs and see if the alert is called

File: app/__tests__/CreateNewCode.test.tsx [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+  it("successfully creates code", async () => {
```
Comment (mdesautel) [inline app/__tests__/CreateNewCode.test.tsx:88]: also, you can test the loading indicator when the code is being generated

**Review by mdesautel — COMMENTED** [review]
Hi, thank you for your PR!
I reviewed all you tests, and they are pretty complete for the screens we have at the moment. I am not sure if the tests for myRent, SharedElements, Subrent, etc. are really relevant now as the screens are not complete and will be largely modified later, so the elements checked for will probably not be present in the final layout. But it is a good thing you still added the tests, to not lose coverage. I mentioned small things you should consider inside the tests, but they are already good and test the majority of cases.
Also, it is good that you modified the codeApprovedScreen to accept both the email and the userId as route params! There was an error before which impacted my tests also, so it is good that you caught it.
And good job on transforming all local alerts in global ones!
I will merge your PR as it looks good to me!

**Comment by userdarius** [comment]
Thanks for the valuable comments! Regarding small components and screens I was informed we still need to maximize our coverage to avoid any small bugs that could arise.

**Review by mdesautel — APPROVED** [review]

**Comment by mdesautel** [comment]
All squashed and merged!
